light-node/src/error.rs

```diff
@@ -1,80 +1,38 @@
 //! Error types
 
-use abscissa_core::error::{BoxError, Context};
-use std::{
-    fmt::{self, Display},
-    io,
-    ops::Deref,
-};
+use std::io;
+use std::net;
+
+/// Error type
+pub type Error = anomaly::Error<Kind>;
 
 /// Kinds of errors
-#[derive(Copy, Clone, Debug, Eq, PartialEq)]
-pub enum ErrorKind {
+#[derive(Clone, Debug, thiserror::Error)]
+pub enum Kind {
+    /// Error when parsing an address.
+    #[error(transparent)]
+    AddrParse(#[from] net::AddrParseError),
+
     /// Error in configuration file
+    #[error("config error")]
     Config,
 
     /// Input/output error
-    Io,
-}
-
-impl ErrorKind {
-    /// Create an error context from this error
-    pub fn context(self, source: impl Into<BoxError>) -> Context<ErrorKind> {
-        Context::new(self, Some(source.into()))
-    }
-}
+    #[error("i/o error: {0}")]
+    Io(String),
```

Comment:
Just out of curiosity: why did you decide to wrap the underlying error as a string here (as opposed to using the `#[from]` directive)?

Reply:
Because `io::Error` doesn't implement `Clone`. Which we mind not need anymore on our `Kind` as I refrained from returning our custom errors in the rpc server, as I wasn't able to get mapping to the jsonrpc error to work.

Reply:
One way to work around the lack of `Clone` on `io::Error` is to use the `anomaly::Context` wrapper (as shown in the [docs](https://docs.rs/anomaly/0.2.0/anomaly/index.html#usage)), which would look something like that:

```diff
diff --git a/light-node/src/error.rs b/light-node/src/error.rs
index 094799b..7162c03 100644
--- a/light-node/src/error.rs
+++ b/light-node/src/error.rs
@@ -1,8 +1,9 @@
 //! Error types
 
-use std::io;
 use std::net;
 
+use anomaly::{BoxError, Context};
+
 /// Error type
 pub type Error = anomaly::Error<Kind>;
 
@@ -18,12 +19,15 @@ pub enum Kind {
     Config,
 
     /// Input/output error
-    #[error("i/o error: {0}")]
-    Io(String),
+    #[error("i/o error")]
+    Io,
 }
 
-impl From<io::Error> for Kind {
-    fn from(err: io::Error) -> Self {
-        Self::Io(format!("{}", err))
+impl Kind {
+    /// Add additional context (i.e. include a source error and capture a backtrace).
+    /// You can convert the resulting `Context` into an `Error` by calling `.into()`.
+    pub fn context(self, source: impl Into<BoxError>) -> Context<Self> {
+        Context::new(self, Some(source.into()))
     }
 }
+
diff --git a/light-node/src/rpc.rs b/light-node/src/rpc.rs
index 544dcb4..f6b190f 100644
--- a/light-node/src/rpc.rs
+++ b/light-node/src/rpc.rs
@@ -25,7 +25,7 @@ where
             AccessControlAllowOrigin::Any,
         ]))
         .start_http(&addr.parse().map_err(error::Kind::from)?)
-        .map_err(error::Kind::from)?;
+        .map_err(|e| error::Kind::Io.context(e))?;
 
     srv.wait();
```

Reply:
PR at #413

Reply:
Yeah was looking at the `Context` as we also use it in the light-client crate. You do loose the `From` semantics with it, correct? As in you have to know which variant to construct when mapping.

Reply:
> You do loose the From semantics with it, correct? As in you have to know which variant to construct when mapping.

I would say that you just have to be more explicit, rather than let the compiler pick the appropriate `From` instance based on the (static) type of the underlying error, which in this case is a `io::Error`. So you don't really lose anything aside from the `From` instance, which in this case is not appropriate for `io::Error` because of the lack of `Clone` bound. If `io::Error` was `Clone`, then you would use `#[from]` like you did for `net::AddrParseError`.